### PR TITLE
1 - Vni phase 2 CRUD, floating ip, reserved ip, flow logs

### PR DESCRIPTION
**Pull request by ujjwal-ibm**
## Changes in datasource :

- [x] virtual_network_interface
  - [x]   code
  - [x]   test
  - [x]   doc

- [x] virtual_network_interface_floating_ip
  - [x]   code
  - [x]   test
  - [x]   doc
- [x] virtual_network_interface_floating_ips
  - [x]   code
  - [x]   test
  - [x]   doc
- [x] virtual_network_interface_ip
  - [x]   code
  - [x]   test
  - [x]   doc

- [x] virtual_network_interface_ips
  - [x]   code
  - [x]   test
  - [x]   doc
- [x] virtual_network_interfaces
  - [x]   code
  - [x]   test
  - [x]   doc

## Changes in resource :

- [x] flow_log
  - [x]   code
  - [ ]   test
  - [ ]   doc

- [x] security_group
  - [x]   code
  - [ ]   test
  - [ ]   doc

- [x] subnet_reserved_ip
  - [x]   code
  - [ ]   test
  - [ ]   doc

- [x] virtual_network_interface
  - [x]   code
  - [x]   test
  - [x]   doc

- [x] virtual_network_interface_floating_ip
  - [x]   code
  - [x]   test
  - [x]   doc

- [x] virtual_network_interface_ip
  - [x]   code
  - [x]   test
  - [x]   doc

[] Develop resource/commands for VNI CRUD apis
[] Develop VNI floating ips CRUD apis
[] Develop VNI reserved ips CRUD apis
[] Revisit VNI prototype and add additional properties to the schema/commands
[] Develop secondary ips in file share mount target
[] develop vni ID and auto_delete in mount target

----

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceBasic (141.15s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     143.277s
...
```

```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceIPBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceIPBasic (142.16s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     144.214s
```
```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceFloatingIPBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceFloatingIPBasic (164.57s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     166.874s
```
```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceFloatingIPsDataSourceBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceFloatingIPsDataSourceBasic (166.14s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     168.242s
```
```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceFloatingIPDataSourceBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceFloatingIPDataSourceBasic (165.28s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     167.334s
```

```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceIpDataSourceBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceIpDataSourceBasic (153.44s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     155.622s
```

```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceIpsDataSourceBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceIpsDataSourceBasic (147.24s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     149.332s
```

```
=== RUN   TestAccIBMISFlowLog_vni
--- PASS: TestAccIBMISFlowLog_vni (159.36s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     161.166s
````

```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceDataSourceVniBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfaceDataSourceVniBasic (46.92s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     49.156s
```

```
=== RUN   TestAccIBMIsVirtualNetworkInterfacesDataSourceVniBasic
--- PASS: TestAccIBMIsVirtualNetworkInterfacesDataSourceVniBasic (42.89s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     44.980s
```

```
=== RUN   TestAccIBMIsVirtualNetworkInterfaceAllArgs
--- PASS: TestAccIBMIsVirtualNetworkInterfaceAllArgs (250.12s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc     252.792s
```

### CODE REVIEW

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+									},
+								},
+							},
+						},
+						"href": &schema.Schema{
+							Type:        schema.TypeString,
+							Required:    true,
+							Description: "The URL for this reserved IP.",
+						},
+						"id": &schema.Schema{
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:77]: ips.id should be mutually exclusive with ips.name and ips.address.

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go [hunk 2]
```diff
@@ -0,0 +1,1115 @@
...
+						},
+						"name": &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The name for this resource group.",
+						},
+					},
+				},
+			},
+			"security_groups": &schema.Schema{
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:176]: To be consistent with other resource, this should be a typeset that takes set of  string of IDs

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface_floating_ip_test.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+		resource "ibm_is_floating_ip" "fip" {
+			name = "%s"
+			zone = "%s"
+		}
+		resource "ibm_is_virtual_network_interface" "vni" {
+			allow_ip_spoofing = %s
+			auto_delete = %s
+			enable_infrastructure_nat = %s
+			ips {
+				address = "192.168.3.4"
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_virtual_network_interface_floating_ip_test.go:87]: address and id should be mutually exclusive

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go [hunk 4]
```diff
@@ -0,0 +1,1035 @@
...
+						},
+						"resource_type": &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The resource type.",
+						},
+					},
+				},
+			},
+			"resource_group": &schema.Schema{
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:151]: For resources, "resource_group" is schema.TypeString in all existing resources

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go [hunk 5]
```diff
@@ -0,0 +1,1035 @@
...
+				},
+			},
+			"subnet": &schema.Schema{
+				Type:        schema.TypeList,
+				MaxItems:    1,
+				Optional:    true,
+				Description: "The associated subnet.",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"crn": &schema.Schema{
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:227]: "subnet" should be a typestring for Id like it is in existing resources

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go [hunk 6]
```diff
@@ -0,0 +1,1035 @@
...
+						},
+						"resource_type": &schema.Schema{
+							Type:        schema.TypeString,
+							Optional:    true,
+							Description: "The resource type.",
+						},
+					},
+				},
+			},
+			"vpc": &schema.Schema{
```
Comment (deepaksibm) [inline ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:344]: Same as above

**Comment by astha-jain** [comment]
https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/is_virtual_network_interface - Is missing attributes like enable_infrastructure_nat and allow_ip_spoofing. Please add.

**Comment by astha-jain** [comment]
Datasource filter by resource_group.id - https://pages.github.ibm.com/vpc/vpc-spec-artifacts/branch/master/swagger-ui.html?feature=is-virtual-network-interface-phase-2#/

File: ibm/service/vpc/resource_ibm_is_bare_metal_server.go [hunk 7]
```diff
@@ -696,7 +696,8 @@ func resourceIBMISBareMetalServerCreate(context context.Context, d *schema.Resou
 	if err != nil {
```
Comment (astha-jain) [inline ibm/service/vpc/resource_ibm_is_bare_metal_server.go:696]: Does changes in this file belong to this feature

**Comment by astha-jain** [comment]
datasource for reserved ips - Should be covered in this PR Right? 

[vni-crud-branch][parent branch vni-1]
Generate SDK and keep it updated

- [ ]  Develop resource/commands for VNI CRUD apis
- [ ]  Develop VNI floating ips CRUD apis
- [ ]  Develop VNI reserved ips CRUD apis
- [ ]  Revisit VNI prototype and add additional properties to the schema/commands
- [ ]  Develop secondary ips in file share mount target
- [ ]  develop vni ID and auto_delete in mount target

File: ibm/provider/provider.go [hunk 8]
```diff
@@ -1036,6 +1036,7 @@ func Provider() *schema.Provider {
 			"ibm_is_subnet_routing_table_attachment":        vpc.ResourceIBMISSubnetRoutingTableAttachment(),
 			"ibm_is_ssh_key":                                vpc.ResourceIBMISSSHKey(),
 			"ibm_is_snapshot":                               vpc.ResourceIBMSnapshot(),
+			"ibm_is_virtual_network_interface":              vpc.ResourceIBMIsVirtualNetworkInterface(),
 			"ibm_is_volume":                                 vpc.ResourceIBMISVolume(),
```
Comment (astha-jain) [inline ibm/provider/provider.go:1040]: datasource for floating ips?

Comment (ujjwal-ibm) [reply ibm/provider/provider.go:1070]: added

File: ibm/service/vpc/resource_ibm_is_bare_metal_server.go (hunk 7 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_bare_metal_server.go:696]: no, sdk changes

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go (hunk 1 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_virtual_network_interface.go]: code changed

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go (hunk 2 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_virtual_network_interface.go]: done

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface_floating_ip_test.go (hunk 3 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_virtual_network_interface_floating_ip_test.go]: code changed

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go (hunk 4 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:178]: code changed

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go (hunk 5 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_virtual_network_interface.go]: code changed

File: ibm/service/vpc/resource_ibm_is_virtual_network_interface.go (hunk 6 above)
Comment (ujjwal-ibm) [reply ibm/service/vpc/resource_ibm_is_virtual_network_interface.go:270]: done

File: website/docs/d/is_flow_logs.html.markdown [hunk 9]
```diff
@@ -48,7 +48,7 @@ Review the argument references that you can specify for your data source.
 - `vpc` - (String) The ID of the VPC this flow log collector resides in
 - `resource_group` - (String) The ID of the Resource group this flow log collector belongs to
 - `target` - (String) The ID of the target this collector is collecting flow logs for.
-- `target_resource_type` - (String) The target resource type for this flow log collector. Available options are `instance`, `network_interface`, `subnet`, `vpc`
+- `target_resource_type` - (String) The target resource type for this flow log collector. Available options are `instance`, `network_interface`, `subnet`, `vpc`, `virtual_network_interface`
```
Comment (sreekarbvibm) [inline website/docs/d/is_flow_logs.html.markdown:51]: is network_attachment required here ?

File: website/docs/d/is_virtual_network_interface_ip.html.markdown [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+- `reserved_ip` - (Required, Forces new resource, String) The reserved IP identifier.
+- `virtual_network_interface` - (Required, Forces new resource, String)  The virtual network interface identifier.
+
+## Attribute Reference
+
+After your data source is created, you can read values from the following attributes.
+
+- `address` - (String) The IP address.If the address has not yet been selected, the value will be `0.0.0.0`.This property may add support for IPv6 addresses in the future. When processing a value in this property, verify that the address is in an expected format. If it is not, log an error. Optionally halt processing and surface the error, or bypass the resource on which the unexpected IP address format was encountered.
+- `href` - (String) The URL for this reserved IP.
```
Comment (sreekarbvibm) [inline website/docs/d/is_virtual_network_interface_ip.html.markdown:34]: deleted is missing

File: website/docs/r/is_virtual_network_interface_ip.html.markdown [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+- `reserved_ip` - (Required, Forces new resource, String) The reserved IP identifier.
+- `virtual_network_interface` - (Required, Forces new resource, String)  The virtual network interface identifier.
+
+## Attribute Reference
+
+After your resource is created, you can read values from the listed arguments and the following attributes.
+
+- `address` - (String) The IP address.If the address has not yet been selected, the value will be `0.0.0.0`.This property may add support for IPv6 addresses in the future. When processing a value in this property, verify that the address is in an expected format. If it is not, log an error. Optionally halt processing and surface the error, or bypass the resource on which the unexpected IP address format was encountered.
+- `href` - (String) The URL for this reserved IP.
```
Comment (sreekarbvibm) [inline website/docs/r/is_virtual_network_interface_ip.html.markdown:34]: deleted is missing

**Review by sreekarbvibm — CHANGES_REQUESTED** [review]

File: website/docs/d/is_flow_logs.html.markdown (hunk 9 above)
Comment (sumitkumartiwari) [reply website/docs/d/is_flow_logs.html.markdown:51]: Check code and docs..

File: website/docs/r/is_virtual_network_interface_ip.html.markdown (hunk 11 above)
Comment (ujjwal-ibm) [reply website/docs/r/is_virtual_network_interface_ip.html.markdown:34]: added

File: website/docs/d/is_virtual_network_interface_ips.html.markdown [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+## Attribute Reference
+
+After your data source is created, you can read values from the following attributes.
+
+- `id` - The unique identifier of the virtual network interface reserved IP Collection.
+
+- `reserved_ips` - (List) Collection of reserved IPs in this subnet.
+	Nested schema for **reserved_ips**:
+	- `address` - (String) The IP address.If the address has not yet been selected, the value will be `0.0.0.0`.This property may add support for IPv6 addresses in the future. When processing a value in this property, verify that the address is in an expected format. If it is not, log an error. Optionally halt processing and surface the error, or bypass the resource on which the unexpected IP address format was encountered.
+	- `href` - (String) The URL for this reserved IP.
```
[the remark this answers is not in the document]
Comment (ujjwal-ibm) [reply website/docs/d/is_virtual_network_interface_ips.html.markdown:36]: added

File: website/docs/d/is_virtual_network_interface_ip.html.markdown (hunk 10 above)
Comment (ujjwal-ibm) [reply website/docs/d/is_virtual_network_interface_ip.html.markdown:34]: added

File: website/docs/d/is_flow_logs.html.markdown (hunk 9 above)
Comment (ujjwal-ibm) [reply website/docs/d/is_flow_logs.html.markdown:51]: added

**Pull request merged**